miniapps/blocksolvers/div_free_solver.hpp

```diff
@@ -0,0 +1,252 @@
+#include "mfem.hpp"
+#include <memory>
+
+using namespace std;
+using namespace mfem;
+
+/// Parameters for iterative solver
+struct IterSolveParameters
+{
+   int print_level = 0;
+   int max_iter = 500;
+   double abs_tol = 1e-12;
+   double rel_tol = 1e-9;
+};
+
+/// Parameters for the divergence free solver
+struct DFSParameters : IterSolveParameters
+{
+   bool verbose = false;
+   bool B_has_nullity_one = false;   // whether B has a 1D nullspace
+   bool coupled_solve = false;       // whether to solve all unknowns together
+   IterSolveParameters BBT_solve_param;
+};
+
+/// Data for the divergence free solver
+struct DFSData
+{
+   Array<OperatorPtr> agg_hdivdof;    // agglomerates to H(div) dofs table
+   Array<OperatorPtr> agg_l2dof;      // agglomerates to L2 dofs table
+   Array<OperatorPtr> P_hdiv;         // Interpolation matrix for H(div) space
+   Array<OperatorPtr> P_l2;           // Interpolation matrix for L2 space
+   Array<OperatorPtr> P_hcurl;        // Interpolation for kernel space of div
+   Array<OperatorPtr> Q_l2;           // Q_l2[l] = (W_{l+1})^{-1} P_l2[l]^T W_l
+   Array<int> coarsest_ess_hdivdofs;  // coarsest level essential H(div) dofs
+   Array<OperatorPtr> C;              // discrete curl: ND -> RT
+   DFSParameters param;
+};
+
+/// For collecting data needed for the divergence free solver
+class DFSDataCollector
```

Comment:
To avoid confusion with `DFSData` (and I think describe better what this actually does), I would rename this as `DFSSpaces`.